src/server/pps/server/api_server.go

```diff
@@ -1004,42 +1001,41 @@ func (a *apiServer) GetLogs(request *pps.GetLogsRequest, apiGetLogsServer pps.AP
 	// and channels are read into the output server in a stable order.
 	// (sort the pods to make sure that the order of log lines is stable)
 	sort.Sort(podSlice(pods))
-	logChs := make([]chan *pps.LogMessage, len(pods))
-	errCh := make(chan error, 1)
-	done := make(chan struct{})
-	defer close(done)
-	for i := 0; i < len(pods); i++ {
-		logChs[i] = make(chan *pps.LogMessage)
-	}
-	for i, pod := range pods {
-		i, pod := i, pod
-		go func() {
-			defer close(logChs[i]) // Main thread reads from here, so must close
-			// Get full set of logs from pod i
-			err := backoff.Retry(func() error {
-				result := a.kubeClient.CoreV1().Pods(a.namespace).GetLogs(
+	logCh := make(chan *pps.LogMessage)
+	var eg errgroup.Group
+	var mu sync.Mutex
+	eg.Go(func() error {
+		for _, pod := range pods {
+			pod := pod
+			if !request.Follow {
+				mu.Lock()
+			}
```

Comment:
So let me make sure I understand this bit. You're putting the `mu.Lock()` here outside the goro so that the subsequent pod will block on the lock to get its logs. This guarantees that the logs come back in a particular order. But we want to keep the goro around because of a) the follow case and b) convenience / use of the err group

Reply:
Yeah, we do it outside the goro because if we did it inside we could spawn 2 goros and have the second one get the lock before the first since go offers no guarantees about goro spawn order. This code is arguably a little weird, because in the non-follow case `eg.Go` call immediately below this comment could actually be removed and you could just call the function inline (at which point you don't need a function) and get the same semantics. However I think this is overall fine because it allows us to use the same code for both paths with complicating things too much which is why I did it this way.

Reply:
Yup :+1: makes sense

Reply:
One more clarification on this: `GetLogs` will have slightly higher latency with this implementation than the previous one. This is because, in the non-follow case, it requests each pod's logs serially, sending those logs to the caller before it moves on to the next pod. In the previous version it would request logs from all pods concurrently, and then send them 1 at a time. This implementation has the advantage that it doesn't use as much memory as the previous version which I think is an overall win since logs can actually be quite large. Furthermore the latency should actually be similar in terms of time to first byte, just slower in terms of time to last byte, so I suspect in most cases users won't notice.